pocs/linux/kernelctf/CVE-2024-1085_cos/docs/exploit.md

```diff
@@ -0,0 +1,164 @@
+#define nft_deactivate_next(__net, __obj)			\
+        (__obj)->genmask = nft_genmask_next(__net)
+
+/* After committing the ruleset, clear the stale generation bit. */
+#define nft_clear(__net, __obj)					\
+	(__obj)->genmask &= ~nft_genmask_next(__net)
+#define nft_active_genmask(__obj, __genmask)			\
+	!((__obj)->genmask & __genmask)
+```
+
+When an object is deleted, developers usually call `nft_deactivate_next` to modify its `genmask`. Similarly, when developers need to confirm which objects have not been deleted, they need to use `nft_is_active_next` to check to avoid double free and other problems.
+
+## Cause anaylysis
+
+In function `nft_setelem_catchall_deactivate`, it checks if an elem is active by this : 
+
+```c
+...
+list_for_each_entry(catchall, &set->catchall_list, list) {
+		ext = nft_set_elem_ext(set, catchall->elem);
+		if (!nft_is_active(net, ext))
+			continue;
+...
+```
+but it should use function `nft_is_active_next`. This vulnerability makes it possible to free a catchall->elem twice.
+
+
+## Triggering the vulnerability
+
+It's easy to trigger it by following this steps:
+
+- Create a pipapo set A, and a catchall set element B in pipapo set A.
+- Delete element B. 
+- Delete element B again. 
+
+By the way, we need to send the command of step 2 and step 3 together because we need to avoid set element B being actually released before our step 3.
+
+## Exploit it
+
+### Target object caches
+Because the `setelem` object size we plan to use is between 0xc0-0x100, our target object cache is `kmalloc-256`
+
+### Exploit detail
+I exploit CVE-2024-1085 by following steps:
+
+- 1. Create a pipapo set `A`, and a catchall set element `B` in it.
+- 2. Trigger the vulnerability by following messages:
+   
+   ```c
+    msg_list[0] = del_setelem_msg(table, pipapo_set, NULL, 0, NULL, 0, 1);//delete the catchall set element first time
+    msg_list[1] = del_table_msg(test_table);//kfree another heap to avoid crash
+    msg_list[2] = del_setelem_msg(table, pipapo_set, NULL, 0, NULL, 0, 1);//delete the catchall set element second time
+    send_msg_list(socket, msg_list, 3);
+   ```
+   	After this we kfree the catchall set element `B` twice. This makes it possiable that we alloc the heap back twice.
+- 3. Try to alloc the heap of the catchall set element `B` back by creating `nft_table` with `NFTA_TABLE_USERDATA`. Keep allocing heap, and each time you alloc a heap, check whether the heap has been alloced for(confirmed by whether the memory of the already created heap has been modified). After this step, We will find two `nft_table` with the same `udata`. We assume that the two `nft_tables` are `nft_table C` and `nft_table D`.
+- 4. Delete `nft_table C`.
+- 5. Spray heap to get the heap of `nft_table C->udata`
+back. I spray heap by creating set element with `NFTA_SET_ELEM_EXPR` because I want to leak the `ops` pointer of the `nft_expr`.
+- 6. Dump `nft_table D`. Now we leak `nft_last_ops`.

```

Comment:
Mention that `nft_table D`'s udata contains the sprayed `nft_expr` object, so we can leak it's `ops` field which contains `nft_last_ops`.